mods/d2k/maps/harkonnen-09a/harkonnen09a-AI.lua

```diff
@@ -46,35 +93,101 @@ ActivateAI = function()
-	ProduceUnits(CorrinoMain, CLightFactory1, delay, vehilcesToBuild, AttackGroupSize[Difficulty], attackThresholdSize)
-	ProduceUnits(CorrinoMain, CHeavyFactory1, delay, tanksToBuildCorrinoMain, AttackGroupSize[Difficulty], attackThresholdSize)
-	ProduceUnits(CorrinoMain, CStarport, delay, unitsToBuyCorrino, AttackGroupSize[Difficulty], attackThresholdSize)
-
-	ProduceUnits(CorrinoSmall, CBarracks2, delay, infantryToBuild, AttackGroupSize[Difficulty], attackThresholdSize)
-	ProduceUnits(CorrinoSmall, CLightFactory2, delay, vehilcesToBuild, AttackGroupSize[Difficulty], attackThresholdSize)
-	ProduceUnits(CorrinoSmall, CHeavyFactory2, delay, tanksToBuildCorrinoSmall, AttackGroupSize[Difficulty], attackThresholdSize)
+	Trigger.AfterDelay(InitialProductionDelay["CorrinoMain"][Difficulty], function ()
+		ProduceUnits(CorrinoMain, CBarracks1, delay, infantryToBuildCorrinoMain, AttackGroupSize[Difficulty], attackThresholdSize)
+		ProduceUnits(CorrinoMain, CLightFactory1, delay, vehilcesToBuild, AttackGroupSize[Difficulty], attackThresholdSize)
+		ProduceUnits(CorrinoMain, CHeavyFactory1, delay, tanksToBuildCorrinoMain, AttackGroupSize[Difficulty], attackThresholdSize)
+		ProduceUnits(CorrinoMain, CStarport, delay, unitsToBuyCorrino, AttackGroupSize[Difficulty], attackThresholdSize)
+	end)
+	Trigger.AfterDelay(InitialProductionDelay["CorrinoSmall"][Difficulty], function ()
+		ProduceUnits(CorrinoSmall, CBarracks2, delay, infantryToBuild, AttackGroupSize[Difficulty], attackThresholdSize)
+		ProduceUnits(CorrinoSmall, CLightFactory2, delay, vehilcesToBuild, AttackGroupSize[Difficulty], attackThresholdSize)
+		ProduceUnits(CorrinoSmall, CHeavyFactory2, delay, tanksToBuildCorrinoSmall, AttackGroupSize[Difficulty], attackThresholdSize)
+	end)
+	Trigger.OnProduction(CHeavyFactory2, function (producer, produced)
+		if produced.Type == "combat_tank_h" and producer.Owner.IsBot then
+			AICrushLogic(produced, producer.Owner)
+		end
+	end)
+	Trigger.OnProduction(CHeavyFactory1, function (producer, produced)
+		if produced.Type == "combat_tank_h" and producer.Owner.IsBot then
+			AICrushLogic(produced, producer.Owner)
+		end
+	end)
+	Trigger.OnProduction(CStarport, function (producer, produced)
+		if produced.Type == "combat_tank_h.starport" and producer.Owner.IsBot then
+			AICrushLogic(produced, producer.Owner)
+		end
+	end)
+	Trigger.OnProduction(AHeavyFactory1, function (producer, produced)
+		if produced.Type == "combat_tank_a" and producer.Owner.IsBot then
+			AICrushLogic(produced, producer.Owner)
+		end
+	end)
+	Trigger.OnProduction(AHeavyFactory2, function (producer, produced)
+		if produced.Type == "combat_tank_a" and producer.Owner.IsBot then
+			AICrushLogic(produced, producer.Owner)
+		end
+	end)
+	Trigger.OnProduction(AStarport, function (producer, produced)
+		if produced.Type == "combat_tank_a.starport" and producer.Owner.IsBot then
+			AICrushLogic(produced, producer.Owner)
+		end
+	end)
```

Comment:
```suggestion
	local crusherTypes = { "combat_tank_a", "combat_tank_a.starport", "combat_tank_h", "combat_tank_h.starport" }
	local crusherFactories = { AStarport, AHeavyFactory1, AHeavyFactory2, CHeavyFactory1, CHeavyFactory2, CStarport }

Utils.Do(crusherFactories, function(factory)
	Trigger.OnProduction(factory, function(producer, produced)
		if not producer.Owner.IsBot then
			return
		end

		local crusher = Utils.Any(crusherTypes, function(ct) return produced.Type == ct end)

		if crusher then
			AICrushLogic(produced, producer.Owner)
		end
	end)
end)
```
I figure this should make things much easier to read, edit, and apply to other missions.

Reply:
Probably It makes even more sense to move this as function into campaign.lua .